File: deploy/testing/benchmark/benchmark.py
#!/usr/bin/env python3
"""
=============================================================================
GOQUEUE PYTHON BENCHMARK SUITE
=============================================================================

This script tests GoQueue throughput and latency using Python.

USAGE:
    # Run benchmark against local GoQueue
    GOQUEUE_URL=http://localhost:8080 python3 benchmark.py

    # Run against EKS cluster
    GOQUEUE_URL=http://<load-balancer-url>:8080 python3 benchmark.py

REQUIREMENTS:
    pip install requests

=============================================================================
"""

import json
import os
import random
import statistics
import string
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from typing import List, Dict, Any
import requests

# =============================================================================
# CONFIGURATION
# =============================================================================

GOQUEUE_URL = os.environ.get("GOQUEUE_URL", "http://localhost:8080")

# =============================================================================
# DATA CLASSES
# =============================================================================

@dataclass
class BenchmarkResult:
    name: str
    total_messages: int
    duration_ms: float
    throughput: float  # msgs/sec
    avg_latency_ms: float
    p50_latency_ms: float
    p95_latency_ms: float
    p99_latency_ms: float


# =============================================================================
# HTTP CLIENT
# =============================================================================

class GoQueueClient:
    def __init__(self, base_url: str):
        self.base_url = base_url
        self.session = requests.Session()
        # Connection pooling
        adapter = requests.adapters.HTTPAdapter(
            pool_connections=100,
            pool_maxsize=100,
        )
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

    def create_topic(self, name: str, partitions: int) -> None:
        # GoQueue API uses /topics (not /api/v1/topics)
        response = self.session.post(
            f"{self.base_url}/topics",
            json={"name": name, "num_partitions": partitions},
        )
        if response.status_code not in (201, 409):
            raise Exception(f"Failed to create topic: {response.status_code}")

    def delete_topic(self, name: str) -> None:
        self.session.delete(f"{self.base_url}/topics/{name}")

    def publish(self, topic: str, messages: List[Dict[str, Any]]) -> None:
        response = self.session.post(
            f"{self.base_url}/topics/{topic}/messages",
            json={"messages": messages},
        )
        if response.status_code not in (200, 201):
            raise Exception(f"Publish failed: {response.status_code}")

    def join_group(self, group_id: str, client_id: str, topics: List[str]) -> str:
        response = self.session.post(
            f"{self.base_url}/groups/{group_id}/join",
            json={"client_id": client_id, "topics": topics},
        )
        if response.status_code not in (200, 201):
            raise Exception(f"Join group failed: {response.status_code}")
        return response.json()["member_id"]

    def poll(self, group_id: str, member_id: str) -> Dict[str, Any]:
        # Poll uses GET /groups/{group}/poll?member_id=xxx
        response = self.session.get(
            f"{self.base_url}/groups/{group_id}/poll",
            params={"member_id": member_id, "timeout": "5s"}
        )
        if response.status_code != 200:
            raise Exception(f"Poll failed: {response.status_code} - {response.text}")
        return response.json()

    def health_check(self) -> bool:
        try:
            response = self.session.get(f"{self.base_url}/health")
            return response.status_code == 200
        except:
            return False


# =============================================================================
# HELPERS
# =============================================================================

def generate_message(size: int) -> Dict[str, str]:
    """Generate a random message of specified size."""
    value = "".join(random.choices(string.ascii_letters + string.digits, k=size))
    return {"value": value}


def generate_messages(count: int, size: int) -> List[Dict[str, str]]:
    """Generate a batch of messages."""
    return [generate_message(size) for _ in range(count)]


def percentile(values: List[float], p: float) -> float:
    """Calculate the p-th percentile of values."""
    if not values:
        return 0.0
    sorted_values = sorted(values)
    idx = int((p / 100) * len(sorted_values))
    return sorted_values[min(idx, len(sorted_values) - 1)]


# =============================================================================
# BENCHMARKS
# =============================================================================

def benchmark_publish_throughput(client: GoQueueClient) -> BenchmarkResult:
    """Benchmark batch publish throughput."""
    print("\n📊 Running: Publish Throughput Benchmark")
    print("   Testing batch publish with 100 messages x 1KB each...")

    topic_name = f"bench-py-{int(time.time() * 1000)}"
    client.create_topic(topic_name, 6)

    batch_size = 100
    message_size = 1024
    iterations = 100
    messages = generate_messages(batch_size, message_size)
    latencies = []

    start_time = time.time()

    for _ in range(iterations):
        batch_start = time.time()
        client.publish(topic_name, messages)
        latencies.append((time.time() - batch_start) * 1000)

    duration_ms = (time.time() - start_time) * 1000
    total_messages = iterations * batch_size

    client.delete_topic(topic_name)

    return BenchmarkResult(
        name="Publish Throughput (batch=100, size=1KB)",
        total_messages=total_messages,
        duration_ms=duration_ms,
        throughput=(total_messages / duration_ms) * 1000,
        avg_latency_ms=statistics.mean(latencies),
        p50_latency_ms=percentile(latencies, 50),
        p95_latency_ms=percentile(latencies, 95),
        p99_latency_ms=percentile(latencies, 99),
    )


def benchmark_concurrent_publish(client: GoQueueClient) -> BenchmarkResult:
    """Benchmark concurrent publish throughput."""
    print("\n📊 Running: Concurrent Publish Benchmark")
    print("   Testing 8 concurrent producers...")

    topic_name = f"bench-concurrent-{int(time.time() * 1000)}"
    client.create_topic(topic_name, 6)

    num_producers = 8
    batch_size = 100
    iterations_per_producer = 50
    message_size = 1024
    messages = generate_messages(batch_size, message_size)
    latencies = []

    def producer_task():
        local_latencies = []
        for _ in range(iterations_per_producer):
            batch_start = time.time()
            client.publish(topic_name, messages)
            local_latencies.append((time.time() - batch_start) * 1000)
        return local_latencies

    start_time = time.time()

    with ThreadPoolExecutor(max_workers=num_producers) as executor:
        futures = [executor.submit(producer_task) for _ in range(num_producers)]
        for future in as_completed(futures):
            latencies.extend(future.result())

    duration_ms = (time.time() - start_time) * 1000
    total_messages = num_producers * iterations_per_producer * batch_size

    client.delete_topic(topic_name)

    return BenchmarkResult(
        name=f"Concurrent Publish (producers={num_producers})",
        total_messages=total_messages,
        duration_ms=duration_ms,
        throughput=(total_messages / duration_ms) * 1000,
        avg_latency_ms=statistics.mean(latencies),
        p50_latency_ms=percentile(latencies, 50),
        p95_latency_ms=percentile(latencies, 95),
        p99_latency_ms=percentile(latencies, 99),
    )


def benchmark_end_to_end_latency(client: GoQueueClient) -> BenchmarkResult:
    """Benchmark end-to-end latency."""
    print("\n📊 Running: End-to-End Latency Benchmark")
    print("   Measuring publish → consume latency...")

    topic_name = f"bench-e2e-{int(time.time() * 1000)}"
    group_name = f"bench-group-{int(time.time() * 1000)}"
    client.create_topic(topic_name, 1)

    member_id = client.join_group(group_name, "latency-tester", [topic_name])

    iterations = 200
    latencies = []

    for i in range(iterations):
        message = {"value": json.dumps({"ts": time.time(), "seq": i})}
        
        start_time = time.time()
        client.publish(topic_name, [message])
        client.poll(group_name, member_id)
        latencies.append((time.time() - start_time) * 1000)

    client.delete_topic(topic_name)

    total_duration = sum(latencies)

    return BenchmarkResult(
        name="End-to-End Latency",
        total_messages=iterations,
        duration_ms=total_duration,
        throughput=(iterations / total_duration) * 1000,
        avg_latency_ms=statistics.mean(latencies),
        p50_latency_ms=percentile(latencies, 50),
        p95_latency_ms=percentile(latencies, 95),
        p99_latency_ms=percentile(latencies, 99),
    )


def benchmark_sustained_throughput(client: GoQueueClient) -> BenchmarkResult:
    """Benchmark sustained throughput over time."""
    print("\n📊 Running: Sustained Throughput Benchmark (30s)")
    print("   Running sustained load with 4 producers...")

    topic_name = f"bench-sustained-{int(time.time() * 1000)}"
    client.create_topic(topic_name, 6)

    duration_secs = 30
    num_producers = 4
    batch_size = 100
    message_size = 1024
    messages = generate_messages(batch_size, message_size)

    total_messages = 0
    running = True
    latencies = []
    end_time = time.time() + duration_secs

    def producer_task():
        nonlocal total_messages
        local_latencies = []
        while time.time() < end_time:
            try:
                batch_start = time.time()
                client.publish(topic_name, messages)
                local_latencies.append((time.time() - batch_start) * 1000)
                total_messages += batch_size
            except:
                pass
        return local_latencies

    start_time = time.time()

    with ThreadPoolExecutor(max_workers=num_producers) as executor:
        futures = [executor.submit(producer_task) for _ in range(num_producers)]
        for future in as_completed(futures):
            latencies.extend(future.result())

    actual_duration = (time.time() - start_time) * 1000

    client.delete_topic(topic_name)

    return BenchmarkResult(
        name=f"Sustained Throughput ({duration_secs}s, {num_producers} producers)",
        total_messages=total_messages,
        duration_ms=actual_duration,
        throughput=(total_messages / actual_duration) * 1000 if actual_duration > 0 else 0,
        avg_latency_ms=statistics.mean(latencies) if latencies else 0,
        p50_latency_ms=percentile(latencies, 50),
        p95_latency_ms=percentile(latencies, 95),
        p99_latency_ms=percentile(latencies, 99),
    )


# =============================================================================
# OUTPUT
# =============================================================================

def print_results(results: List[BenchmarkResult]) -> None:
    """Print benchmark results in a formatted table."""
    print("\n" + "═" * 100)
    print("                              GOQUEUE BENCHMARK RESULTS (Python)")
    print("═" * 100)
    print(f"Target: {GOQUEUE_URL}")
    print(f"Date: {time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}")
    print("─" * 100)

    print("\n┌" + "─" * 56 + "┬" + "─" * 12 + "┬" + "─" * 12 + "┬" + "─" * 13 + "┐")
    print("│ Benchmark                                              │ Throughput │ Avg Latency│ P99 Latency │")
    print("│                                                        │ (msgs/sec) │   (ms)     │    (ms)     │")
    print("├" + "─" * 56 + "┼" + "─" * 12 + "┼" + "─" * 12 + "┼" + "─" * 13 + "┤")

    for result in results:
        name = result.name[:54].ljust(54)
        throughput = f"{result.throughput:,.0f}".rjust(10)
        avg_latency = f"{result.avg_latency_ms:,.2f}".rjust(10)
        p99_latency = f"{result.p99_latency_ms:,.2f}".rjust(11)
        print(f"│ {name} │ {throughput} │ {avg_latency} │ {p99_latency} │")

    print("└" + "─" * 56 + "┴" + "─" * 12 + "┴" + "─" * 12 + "┴" + "─" * 13 + "┘")

    # Comparison table
    print("\n" + "─" * 100)
    print("                              INDUSTRY COMPARISON (Reference)")
    print("─" * 100)
    
    last_throughput = f"{results[-1].throughput:,.0f}" if results else "N/A"
    last_p99 = f"{results[-1].p99_latency_ms:,.2f}" if results else "N/A"
    
    print(f"""
┌──────────────┬─────────────────────┬───────────────┬───────────────────────────────────────────────────┐
│ System       │ Throughput (msg/s)  │ P99 Latency   │ Notes                                             │
├──────────────┼─────────────────────┼───────────────┼───────────────────────────────────────────────────┤
│ Apache Kafka │ 1,000,000+          │ 5-50ms        │ With batching, linger.ms, compression             │
│ RabbitMQ     │ 10,000-100,000      │ 1-10ms        │ Persistent messages, prefetch tuned               │
│ AWS SQS      │ 300,000 (3K FIFO)   │ 50-500ms      │ Managed service, HTTP API overhead                │
│ Redis Streams│ 500,000+            │ <1ms          │ In-memory, optional persistence                   │
│ NATS         │ 10,000,000+         │ <1ms          │ At-most-once default, JetStream for durability    │
├──────────────┼─────────────────────┼───────────────┼───────────────────────────────────────────────────┤
│ GoQueue      │ {last_throughput:>17} │ {last_p99:>11}ms │ This benchmark run                                │
└──────────────┴─────────────────────┴───────────────┴───────────────────────────────────────────────────┘
""")


# =============================================================================
# MAIN
# =============================================================================

def main():
    print("╔" + "═" * 100 + "╗")
    print("║" + "                           GOQUEUE BENCHMARK SUITE (Python)                                   ".ljust(100) + "║")
    print("╠" + "═" * 100 + "╣")
    print(f"║ Target URL: {GOQUEUE_URL}".ljust(101) + "║")
    print("╚" + "═" * 100 + "╝")

    client = GoQueueClient(GOQUEUE_URL)

    # Verify connection
    if not client.health_check():
        print(f"\n❌ Failed to connect to GoQueue at {GOQUEUE_URL}")
        print("   Make sure the server is running and accessible.")
        exit(1)
    
    print("\n✅ Connected to GoQueue server")

    results = []

    try:
        results.append(benchmark_publish_throughput(client))
        results.append(benchmark_concurrent_publish(client))
        # NOTE: End-to-end and sustained tests require consumer group heartbeats
        # which are not implemented in this simple benchmark. Skipping for now.
        # results.append(benchmark_end_to_end_latency(client))
        results.append(benchmark_sustained_throughput(client))
    except Exception as e:
        print(f"\n❌ Benchmark failed: {e}")
        exit(1)

    print_results(results)


if __name__ == "__main__":
    main()

File: deploy/testing/benchmark/benchmark.ts
// =============================================================================
// GOQUEUE TYPESCRIPT BENCHMARK SUITE
// =============================================================================
//
// This script tests GoQueue throughput and latency using the TypeScript client.
//
// USAGE:
//   # Install dependencies
//   cd deploy/testing/benchmark
//   npm install
//
//   # Run benchmark against local GoQueue
//   GOQUEUE_URL=http://localhost:8080 npx tsx benchmark.ts
//
//   # Run against EKS cluster
//   GOQUEUE_URL=http://<load-balancer-url>:8080 npx tsx benchmark.ts
//
// =============================================================================

interface BenchmarkResult {
  name: string;
  totalMessages: number;
  durationMs: number;
  throughput: number;
  avgLatencyMs: number;
  p50LatencyMs: number;
  p95LatencyMs: number;
  p99LatencyMs: number;
}

// =============================================================================
// CONFIGURATION
// =============================================================================

const GOQUEUE_URL = process.env.GOQUEUE_URL || 'http://localhost:8080';

// =============================================================================
// HTTP CLIENT
// =============================================================================

async function createTopic(name: string, partitions: number): Promise<void> {
  // GoQueue API uses /topics (not /api/v1/topics)
  const response = await fetch(`${GOQUEUE_URL}/topics`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      name,
      num_partitions: partitions,
    }),
  });

  if (!response.ok && response.status !== 409) {
    throw new Error(`Failed to create topic: ${response.status} ${await response.text()}`);
  }
}

async function deleteTopic(name: string): Promise<void> {
  await fetch(`${GOQUEUE_URL}/topics/${name}`, {
    method: 'DELETE',
  });
}

async function publish(
  topic: string,
  messages: Array<{ value: string; priority?: number }>
): Promise<void> {
  const response = await fetch(`${GOQUEUE_URL}/topics/${topic}/messages`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ messages }),
  });

  if (!response.ok) {
    throw new Error(`Publish failed: ${response.status}`);
  }
}

async function joinGroup(
  groupId: string,
  clientId: string,
  topics: string[]
): Promise<{ member_id: string }> {
  const response = await fetch(`${GOQUEUE_URL}/groups/${groupId}/join`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      client_id: clientId,
      topics,
    }),
  });

  if (!response.ok) {
    throw new Error(`Join group failed: ${response.status}`);
  }

  return response.json();
}

async function poll(groupId: string, memberId: string): Promise<any> {
  // Poll uses GET /groups/{group}/poll?member_id=xxx
  const response = await fetch(
    `${GOQUEUE_URL}/groups/${groupId}/poll?member_id=${memberId}&timeout=5s`,
    { method: 'GET' }
  );

  if (!response.ok) {
    throw new Error(`Poll failed: ${response.status}`);
  }

  return response.json();
}

// =============================================================================
// HELPERS
// =============================================================================

function generateMessage(size: number): { value: string } {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let value = '';
  for (let i = 0; i < size; i++) {
    value += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return { value };
}

function generateMessages(count: number, size: number): Array<{ value: string }> {
  return Array.from({ length: count }, () => generateMessage(size));
}

function percentile(arr: number[], p: number): number {
  const sorted = [...arr].sort((a, b) => a - b);
  const idx = Math.ceil((p / 100) * sorted.length) - 1;
  return sorted[Math.max(0, idx)];
}

function formatNumber(n: number): string {
  return n.toLocaleString('en-US', { maximumFractionDigits: 2 });
}

// =============================================================================
// BENCHMARKS
// =============================================================================

async function benchmarkPublishThroughput(): Promise<BenchmarkResult> {
  console.log('\n📊 Running: Publish Throughput Benchmark');
  console.log('   Testing batch publish with 100 messages x 1KB each...');

  const topicName = `bench-ts-${Date.now()}`;
  await createTopic(topicName, 6);

  const batchSize = 100;
  const messageSize = 1024;
  const iterations = 100;
  const messages = generateMessages(batchSize, messageSize);
  const latencies: number[] = [];

  const startTime = Date.now();

  for (let i = 0; i < iterations; i++) {
    const batchStart = Date.now();
    await publish(topicName, messages);
    latencies.push(Date.now() - batchStart);
  }

  const durationMs = Date.now() - startTime;
  const totalMessages = iterations * batchSize;

  await deleteTopic(topicName);

  return {
    name: 'Publish Throughput (batch=100, size=1KB)',
    totalMessages,
    durationMs,
    throughput: (totalMessages / durationMs) * 1000,
    avgLatencyMs: latencies.reduce((a, b) => a + b, 0) / latencies.length,
    p50LatencyMs: percentile(latencies, 50),
    p95LatencyMs: percentile(latencies, 95),
    p99LatencyMs: percentile(latencies, 99),
  };
}

async function benchmarkConcurrentPublish(): Promise<BenchmarkResult> {
  console.log('\n📊 Running: Concurrent Publish Benchmark');
  console.log('   Testing 8 concurrent producers...');

  const topicName = `bench-concurrent-${Date.now()}`;
  await createTopic(topicName, 6);

  const numProducers = 8;
  const batchSize = 100;
  const iterationsPerProducer = 50;
  const messageSize = 1024;
  const messages = generateMessages(batchSize, messageSize);
  const latencies: number[] = [];

  const startTime = Date.now();

  const producers = Array.from({ length: numProducers }, async () => {
    for (let i = 0; i < iterationsPerProducer; i++) {
      const batchStart = Date.now();
      await publish(topicName, messages);
      latencies.push(Date.now() - batchStart);
    }
  });

  await Promise.all(producers);

  const durationMs = Date.now() - startTime;
  const totalMessages = numProducers * iterationsPerProducer * batchSize;

  await deleteTopic(topicName);

  return {
    name: `Concurrent Publish (producers=${numProducers})`,
    totalMessages,
    durationMs,
    throughput: (totalMessages / durationMs) * 1000,
    avgLatencyMs: latencies.reduce((a, b) => a + b, 0) / latencies.length,
    p50LatencyMs: percentile(latencies, 50),
    p95LatencyMs: percentile(latencies, 95),
    p99LatencyMs: percentile(latencies, 99),
  };
}

async function benchmarkEndToEndLatency(): Promise<BenchmarkResult> {
  console.log('\n📊 Running: End-to-End Latency Benchmark');
  console.log('   Measuring publish → consume latency...');

  const topicName = `bench-e2e-${Date.now()}`;
  const groupName = `bench-group-${Date.now()}`;
  await createTopic(topicName, 1);

  const { member_id } = await joinGroup(groupName, 'latency-tester', [topicName]);

  const iterations = 200;
  const latencies: number[] = [];

  for (let i = 0; i < iterations; i++) {
    const message = { value: JSON.stringify({ ts: Date.now(), seq: i }) };
    
    const startTime = Date.now();
    await publish(topicName, [message]);
    await poll(groupName, member_id);
    latencies.push(Date.now() - startTime);
  }

  await deleteTopic(topicName);

  const totalDuration = latencies.reduce((a, b) => a + b, 0);

  return {
    name: 'End-to-End Latency',
    totalMessages: iterations,
    durationMs: totalDuration,
    throughput: (iterations / totalDuration) * 1000,
    avgLatencyMs: totalDuration / iterations,
    p50LatencyMs: percentile(latencies, 50),
    p95LatencyMs: percentile(latencies, 95),
    p99LatencyMs: percentile(latencies, 99),
  };
}

async function benchmarkSustainedThroughput(): Promise<BenchmarkResult> {
  console.log('\n📊 Running: Sustained Throughput Benchmark (30s)');
  console.log('   Running sustained load with 4 producers...');

  const topicName = `bench-sustained-${Date.now()}`;
  await createTopic(topicName, 6);

  const durationMs = 30000; // 30 seconds
  const numProducers = 4;
  const batchSize = 100;
  const messageSize = 1024;
  const messages = generateMessages(batchSize, messageSize);

  let totalMessages = 0;
  let running = true;
  const latencies: number[] = [];

  const startTime = Date.now();

  const producers = Array.from({ length: numProducers }, async () => {
    while (running) {
      const batchStart = Date.now();
      try {
        await publish(topicName, messages);
        totalMessages += batchSize;
        latencies.push(Date.now() - batchStart);
      } catch (e) {
        // Ignore errors during benchmark
      }
    }
  });

  await new Promise((resolve) => setTimeout(resolve, durationMs));
  running = false;

  await Promise.allSettled(producers);
  await deleteTopic(topicName);

  const actualDuration = Date.now() - startTime;

  return {
    name: `Sustained Throughput (${durationMs / 1000}s, ${numProducers} producers)`,
    totalMessages,
    durationMs: actualDuration,
    throughput: (totalMessages / actualDuration) * 1000,
    avgLatencyMs: latencies.length > 0 ? latencies.reduce((a, b) => a + b, 0) / latencies.length : 0,
    p50LatencyMs: latencies.length > 0 ? percentile(latencies, 50) : 0,
    p95LatencyMs: latencies.length > 0 ? percentile(latencies, 95) : 0,
    p99LatencyMs: latencies.length > 0 ? percentile(latencies, 99) : 0,
  };
}

// =============================================================================
// MAIN
// =============================================================================

function printResults(results: BenchmarkResult[]): void {
  console.log('\n' + '═'.repeat(100));
  console.log('                              GOQUEUE BENCHMARK RESULTS');
  console.log('═'.repeat(100));
  console.log(`Target: ${GOQUEUE_URL}`);
  console.log(`Date: ${new Date().toISOString()}`);
  console.log('─'.repeat(100));

  console.log('\n┌────────────────────────────────────────────────────────┬────────────┬────────────┬─────────────┐');
  console.log('│ Benchmark                                              │ Throughput │ Avg Latency│ P99 Latency │');
  console.log('│                                                        │ (msgs/sec) │   (ms)     │    (ms)     │');
  console.log('├────────────────────────────────────────────────────────┼────────────┼────────────┼─────────────┤');

  for (const result of results) {
    const name = result.name.padEnd(54);
    const throughput = formatNumber(result.throughput).padStart(10);
    const avgLatency = formatNumber(result.avgLatencyMs).padStart(10);
    const p99Latency = formatNumber(result.p99LatencyMs).padStart(11);
    console.log(`│ ${name} │ ${throughput} │ ${avgLatency} │ ${p99Latency} │`);
  }

  console.log('└────────────────────────────────────────────────────────┴────────────┴────────────┴─────────────┘');

  // Comparison table
  console.log('\n' + '─'.repeat(100));
  console.log('                              INDUSTRY COMPARISON (Reference)');
  console.log('─'.repeat(100));
  console.log(`
┌──────────────┬─────────────────────┬───────────────┬───────────────────────────────────────────────────┐
│ System       │ Throughput (msg/s)  │ P99 Latency   │ Notes                                             │
├──────────────┼─────────────────────┼───────────────┼───────────────────────────────────────────────────┤
│ Apache Kafka │ 1,000,000+          │ 5-50ms        │ With batching, linger.ms, compression             │
│ RabbitMQ     │ 10,000-100,000      │ 1-10ms        │ Persistent messages, prefetch tuned               │
│ AWS SQS      │ 300,000 (3K FIFO)   │ 50-500ms      │ Managed service, HTTP API overhead                │
│ Redis Streams│ 500,000+            │ <1ms          │ In-memory, optional persistence                   │
│ NATS         │ 10,000,000+         │ <1ms          │ At-most-once default, JetStream for durability    │
├──────────────┼─────────────────────┼───────────────┼───────────────────────────────────────────────────┤
│ GoQueue      │ ${formatNumber(results[results.length - 1]?.throughput || 0).padStart(17)} │ ${formatNumber(results[results.length - 1]?.p99LatencyMs || 0).padStart(11)}ms │ This benchmark run                                │
└──────────────┴─────────────────────┴───────────────┴───────────────────────────────────────────────────┘
`);
}

async function main(): Promise<void> {
  console.log('╔════════════════════════════════════════════════════════════════════════════════════════════════╗');
  console.log('║                           GOQUEUE BENCHMARK SUITE (TypeScript)                                  ║');
  console.log('╠════════════════════════════════════════════════════════════════════════════════════════════════╣');
  console.log(`║ Target URL: ${GOQUEUE_URL.padEnd(83)}║`);
  console.log('╚════════════════════════════════════════════════════════════════════════════════════════════════╝');

  // Verify connection
  try {
    const healthResponse = await fetch(`${GOQUEUE_URL}/health`);
    if (!healthResponse.ok) {
      throw new Error(`Health check failed: ${healthResponse.status}`);
    }
    console.log('\n✅ Connected to GoQueue server');
  } catch (error) {
    console.error(`\n❌ Failed to connect to GoQueue at ${GOQUEUE_URL}`);
    console.error('   Make sure the server is running and accessible.');
    process.exit(1);
  }

  const results: BenchmarkResult[] = [];

  try {
    results.push(await benchmarkPublishThroughput());
    results.push(await benchmarkConcurrentPublish());
    // NOTE: End-to-end test requires consumer group heartbeats
    // which are not implemented in this simple benchmark. Skipping for now.
    // results.push(await benchmarkEndToEndLatency());
    results.push(await benchmarkSustainedThroughput());
  } catch (error) {
    console.error('\n❌ Benchmark failed:', error);
    process.exit(1);
  }

  printResults(results);
}

main().catch(console.error);
